File: app/src/main/java/com/jerech/tictactoe/app/api/ApiClient.java
package com.jerech.tictactoe.app.api;

import android.content.Context;
import android.content.Intent;

import com.google.gson.GsonBuilder;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import com.jerech.tictactoe.app.R;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by jeremias on 05/10/19.
 */

public class ApiClient {


    public static final String CONTENT_TYPE = "Content-Type: application/json";

    private static Retrofit retrofit = null;


    private static ApiClient self = null;
    private Context context;


    private ApiClient(Context context){
        this.context = context;
    }




    public static Retrofit getClient(Context context, String token) {

        if(self==null){
            self = new ApiClient(context);
        }
        String url = context.getResources().getString(R.string.base_url_api)+context.getResources().getString(R.string.endpoint);

        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder().connectTimeout(2, TimeUnit.MINUTES)
                .addInterceptor(interceptor)
                .build();
        if (retrofit==null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(url)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create(new GsonBuilder().excludeFieldsWithoutExposeAnnotation() .create()))
                    .build();
        }
        return retrofit;
    }


}

File: app/src/main/java/com/jerech/tictactoe/app/api/ApiClient.kt
package com.jerech.tictactoe.app.api;

import android.content.Context;

import com.google.gson.GsonBuilder;
import com.jerech.tictactoe.app.BuildConfig
import com.jerech.tictactoe.app.R
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory


/**
 * Created by jeremias on 05/10/19.
 */

class ApiClient {


    fun getClient(context: Context?) : Retrofit {

        val client = OkHttpClient().newBuilder()
            .addInterceptor(HttpLoggingInterceptor().apply {
                level = if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE
            })
            .build()

        var retrofit = Retrofit.Builder()
            .baseUrl(context?.resources?.getString(R.string.base_url_api) +
                    context?.resources?.getString(R.string.endpoint))
            .client(client)
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        return retrofit
    }


}
